File: storagemanager/src/main/java/com/example/alexeyglushkov/cachemanager/disk/DiskStorage.java
package com.example.alexeyglushkov.cachemanager.disk;

import androidx.annotation.NonNull;

import com.example.alexeyglushkov.cachemanager.StorageEntry;
import com.example.alexeyglushkov.cachemanager.StorageMetadata;
import com.example.alexeyglushkov.cachemanager.Storage;
import com.example.alexeyglushkov.streamlib.codecs.ObjectCodec;
import com.example.alexeyglushkov.streamlib.codecs.Codec;
import com.example.alexeyglushkov.tools.TimeTools;

import androidx.annotation.Nullable;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Created by alexeyglushkov on 26.09.15.
 */

// TODO: check synchronization
public class DiskStorage implements Storage {
    private static String METADATA_PREFIX = "_metadata";

    private @NonNull File directory;
    private @NonNull Codec defaultCodec = new ObjectCodec();

    private @NonNull Map<String, Codec> keySerializerMap = new HashMap<>();
    private @NonNull Map<Class, Codec> classSerializerMap = new HashMap<>();

    // TODO: need to clear this lockMap somewhere
    private @NonNull Map<String, Object> lockMap = new HashMap<>();

    public DiskStorage(@NonNull File directory) {
        this.directory = directory;
    }

    @Override
    public void put(@NonNull String key, @NonNull Object entry, @Nullable StorageMetadata metadata) throws Exception {
        prepareDirectory();
        write(key, entry, (DiskStorageMetadata) metadata);
    }

    private void prepareDirectory() throws Exception {
        if (!directory.exists()) {
            if (!directory.mkdir()) {
                throw new Exception("DiskStorage.prepareDirectory.mkdir: can't create directory " + directory.getAbsolutePath());
            }
        }
    }

    public File getKeyFile(@NonNull String key) {
        return new File(directory.getPath() + File.separator + key);
    }

    @NonNull
    private File getKeyMetadataFile(@NonNull String key) {
        String fileName = key + METADATA_PREFIX;
        return new File(directory.getPath() + File.separator + fileName);
    }

    private boolean isMetadataFile(@NonNull File file) {
        return file.getName().endsWith(METADATA_PREFIX);
    }

    private void write(@NonNull String fileName, @NonNull Object object, @Nullable DiskStorageMetadata metadata) throws Exception {
        String key = getKeyName(fileName);
        Object lockObject = getLockObject(key);

        synchronized (lockObject) {
            writeByKey(key, object, metadata);
        }
    }

    private static final Set<Character> ReservedCharsSet = new HashSet<>(Arrays.asList(
            '/','|','"','?','*','<','>','\\',':','+','[',']'
    ));

    private String getKeyName(@NonNull String fileName) {
        StringBuilder builder = new StringBuilder(fileName);
        for (int i = 0; i < builder.length(); i++) {
            char c = builder.charAt(i);
            if (ReservedCharsSet.contains(c)) {
                builder.replace(i,i+1,"_");
            }
        }

        return builder.toString();
    }

    synchronized private Object getLockObject(@NonNull String key) {
        Object lockObject = lockMap.get(key);
        if (lockObject == null) {
            lockObject = createLockObject(key);
        }
        return lockObject;
    }

    private Object createLockObject(@NonNull String key) {
        Object lockObject = new Object();
        lockMap.put(key, lockObject);
        return lockObject;
    }

    // TODO: need to refactor
    private void writeByKey(@NonNull String key, @NonNull Object object, @Nullable DiskStorageMetadata metadata) throws Exception {
        File file = getKeyFile(key);
        if (!file.exists()) {
            createFile(file);
        }

        try {
            Codec codec = getSerializer(key, object.getClass());
            if (codec == null) {
                throw new Exception("Can't find a serializer for " + object.getClass());
            }

            DiskStorageEntry entry = new DiskStorageEntry(file, object, metadata, codec);
            entry.write();

            if (metadata != null) {
                writeMetadata(metadata, object, key, file);
            }

        } catch (Exception ex) {
            file.delete();
            if (metadata != null && metadata.getFile() != null) {
                metadata.getFile().delete();
            }

            throw ex;
        }
    }

    private void createFile(@NonNull File file) throws IOException {
        boolean isFileCreated = file.createNewFile();
        if (!isFileCreated) {
            throw new IOException("DiskStorage.write() createNewFile create error");
        }
    }

    private Codec getSerializer(@Nullable String key, @Nullable Class cl) {
        Codec codec = key != null ? keySerializerMap.get(key) : null;
        if (codec == null) {
            codec = cl != null ? classSerializerMap.get(cl) : null;
        }
        return codec == null ? defaultCodec : codec;
    }

    private void writeMetadata(@NonNull DiskStorageMetadata metadata, @NonNull Object object, @NonNull String key, @NonNull File file) throws Exception {
        metadata.setFile(getKeyMetadataFile(key));
        metadata.setCreateTime(TimeTools.currentTimeSeconds());
        metadata.calculateSize(file);
        metadata.setEntryClass(object.getClass());
        metadata.write();
    }

    @Override
    @Nullable
    public Object getValue(@NonNull  String key) throws Exception {
        Object result = null;
        DiskStorageEntry entry = (DiskStorageEntry)getEntry(key);

        if (entry != null) {
            result = entry.getObject();
        }

        return result;
    }

    @Override
    @NonNull
    public DiskStorageMetadata createMetadata() {
        return new DiskStorageMetadata();
    }

    @Nullable
    public StorageMetadata getMetadata(@NonNull String key) throws Exception {
        StorageMetadata result = null;
        StorageEntry entry = getEntry(key);
        if (entry != null) {
            result = entry.getMetadata();
        }
        return result;
    }

    @Override
    @Nullable
    public StorageEntry getEntry(@NonNull String fileName) throws Exception {
        StorageEntry entry = null;
        String key = getKeyName(fileName);
        Object lockObject = getLockObject(key);

        synchronized (lockObject) {
            entry = getEntryByKey(key);
        }

        return entry;
    }

    @NonNull
    private StorageEntry getEntryByKey(@NonNull String key) throws Exception {
        DiskStorageEntry entry = null;
        File file = getKeyFile(key);

        if (!file.exists()) {
            throw new FileNotFoundException("DiskStorage.getEntryByKey() exists(): file doesn't exist");
        }

        DiskStorageMetadata metadata = null;
        Codec codec = null;
        File metadataFile = getKeyMetadataFile(key);

        if (metadataFile.exists()) {
            try {
                metadata = DiskStorageMetadata.load(metadataFile);
                if (metadata != null) {
                    codec = getSerializer(key, metadata.getEntryClass());
                } else {
                    codec = defaultCodec;
                }

            } catch (Exception e) {
                codec = defaultCodec;
            }
        } else {
            codec = defaultCodec;
        }

        if (codec == null) {
            throw new Exception("Serializer is null");
        }

        entry = new DiskStorageEntry(file, null, metadata, codec);
        return entry;
    }

    @Override
    public void remove(@NonNull String key) throws Exception {
        Object lockObject = getLockObject(key);
        synchronized (lockObject) {
            StorageEntry entry = getEntry(key);
            if (entry != null) {
                entry.delete();
            }
        }
    }

    @Override
    public List<StorageEntry> getEntries() throws Exception {
        List<StorageEntry> entries = new ArrayList<>();
        if (!directory.exists()) {
            return entries;
        }

        File[] files = directory.listFiles();

        if (files != null) {
            for (File file : files) {
                if (!isMetadataFile(file)) {
                    String key = file.getName();
                    Object lockObject = getLockObject(key);

                    synchronized (lockObject) {
                        StorageEntry entry = getEntryByKey(key);
                        entries.add(entry);
                    }
                }
            }
        }

        return entries;
    }

    public int getEntryCount() {
        File[] files = directory.listFiles();
        int entryCount = 0;

        for (File file : files) {
            if (!isMetadataFile(file)) {
                ++entryCount;
            }
        }

        return entryCount;
    }

    public void removeAll() throws Exception {
        List<StorageEntry> entries = getEntries();
        for (StorageEntry file : entries) {
            DiskStorageEntry diskCacheEntry = (DiskStorageEntry)file;
            String key = diskCacheEntry.getFileName();
            Object lockObject = getLockObject(key);
            synchronized (lockObject) {
                file.delete();
            }
        }

        if (!directory.delete()) {
            throw new Exception("DiskStorage.removeAll() delete(): remove directory error");
        }
    }

    //// Setter

    public void setSerializer(@NonNull Codec codec, @NonNull Class cl) {
        classSerializerMap.put(cl, codec);
    }

    public void setSerializer(@NonNull Codec codec, @NonNull String key) {
        keySerializerMap.put(key, codec);
    }

    public void setDefaultCodec(@NonNull Codec defaultCodec) {
        this.defaultCodec = defaultCodec;
    }

    //// Getter

    @NonNull
    public File getDirectory() {
        return directory;
    }
}

File: storagemanager/src/main/java/com/example/alexeyglushkov/cachemanager/disk/DiskStorage.kt
package com.example.alexeyglushkov.cachemanager.disk

import com.example.alexeyglushkov.cachemanager.Storage
import com.example.alexeyglushkov.cachemanager.StorageEntry
import com.example.alexeyglushkov.cachemanager.StorageMetadata
import com.example.alexeyglushkov.streamlib.codecs.Codec
import com.example.alexeyglushkov.streamlib.codecs.ObjectCodec
import com.example.alexeyglushkov.tools.TimeTools
import java.io.File
import java.io.FileNotFoundException
import java.io.IOException
import java.util.*
import kotlin.collections.ArrayList
import kotlin.collections.HashMap
import kotlin.collections.HashSet

/**
 * Created by alexeyglushkov on 26.09.15.
 */
// TODO: check synchronization
open class DiskStorage(val directory: File) : Storage {
    private var defaultCodec: Codec<*> = ObjectCodec<Any>()
    private val keySerializerMap: MutableMap<String, Codec<*>> = HashMap()
    private val classSerializerMap: MutableMap<Class<*>, Codec<*>> = HashMap()

    // TODO: need to clear this lockMap somewhere
    private val lockMap: MutableMap<String, Any> = HashMap()

    @Throws(Exception::class)
    override fun put(key: String, entry: Any, metadata: StorageMetadata?) {
        prepareDirectory()
        write(key, entry, metadata as DiskStorageMetadata?)
    }

    @Throws(Exception::class)
    private fun prepareDirectory() {
        if (!directory.exists()) {
            if (!directory.mkdir()) {
                throw Exception("DiskStorage.prepareDirectory.mkdir: can't create directory " + directory.absolutePath)
            }
        }
    }

    fun getKeyFile(key: String): File {
        return File(directory.path + File.separator + key)
    }

    private fun getKeyMetadataFile(key: String): File {
        val fileName = key + METADATA_PREFIX
        return File(directory.path + File.separator + fileName)
    }

    private fun isMetadataFile(file: File): Boolean {
        return file.name.endsWith(METADATA_PREFIX)
    }

    @Throws(Exception::class)
    private fun write(fileName: String, `object`: Any, metadata: DiskStorageMetadata?) {
        val key = getKeyName(fileName)
        val lockObject = getLockObject(key)
        synchronized(lockObject!!) { writeByKey(key, `object`, metadata) }
    }

    private fun getKeyName(fileName: String): String {
        val builder = StringBuilder(fileName)
        for (i in 0 until builder.length) {
            val c = builder[i]
            if (ReservedCharsSet.contains(c)) {
                builder.replace(i, i + 1, "_")
            }
        }
        return builder.toString()
    }

    @Synchronized
    private fun getLockObject(key: String): Any {
        var lockObject = lockMap[key]
        if (lockObject == null) {
            lockObject = createLockObject(key)
        }
        return lockObject
    }

    private fun createLockObject(key: String): Any {
        val lockObject = Any()
        lockMap[key] = lockObject
        return lockObject
    }

    // TODO: need to refactor
    @Throws(Exception::class)
    private fun writeByKey(key: String, `object`: Any, metadata: DiskStorageMetadata?) {
        val file = getKeyFile(key)
        if (!file.exists()) {
            createFile(file)
        }
        try {
            val codec = getSerializer(key, `object`.javaClass)
                    //?: throw Exception("Can't find a serializer for " + `object`.javaClass)
            val entry = DiskStorageEntry(file, `object`, metadata, codec)
            entry.write()
            metadata?.let { writeMetadata(it, `object`, key, file) }
        } catch (ex: Exception) {
            file.delete()
            if (metadata != null && metadata.file != null) {
                metadata.file!!.delete()
            }
            throw ex
        }
    }

    @Throws(IOException::class)
    private fun createFile(file: File) {
        val isFileCreated = file.createNewFile()
        if (!isFileCreated) {
            throw IOException("DiskStorage.write() createNewFile create error")
        }
    }

    private fun getSerializer(key: String?, cl: Class<*>?): Codec<*> {
        var codec = if (key != null) keySerializerMap[key] else null
        if (codec == null) {
            codec = if (cl != null) classSerializerMap[cl] else null
        }
        return codec ?: defaultCodec
    }

    @Throws(Exception::class)
    private fun writeMetadata(metadata: DiskStorageMetadata, `object`: Any, key: String, file: File) {
        metadata.file = getKeyMetadataFile(key)
        metadata.createTime = TimeTools.currentTimeSeconds()
        metadata.calculateSize(file)
        metadata.entryClass = `object`.javaClass
        metadata.write()
    }

    @Throws(Exception::class)
    override fun getValue(key: String): Any? {
        var result: Any? = null
        val entry = getEntry(key) as DiskStorageEntry?
        if (entry != null) {
            result = entry.getObject()
        }
        return result
    }

    override fun createMetadata(): DiskStorageMetadata {
        return DiskStorageMetadata()
    }

    @Throws(Exception::class)
    override fun getMetadata(key: String): StorageMetadata? {
        var result: StorageMetadata? = null
        val entry = getEntry(key)
        if (entry != null) {
            result = entry.metadata
        }
        return result
    }

    @Throws(Exception::class)
    override fun getEntry(fileName: String): StorageEntry? {
        var entry: StorageEntry? = null
        val key = getKeyName(fileName)

        try {
            val lockObject = getLockObject(key)
            synchronized(lockObject) { entry = getEntryByKey(key) }
        } catch (ex: Exception) {
        }
        return entry
    }

    @Throws(Exception::class)
    private fun getEntryByKey(key: String): StorageEntry {
        var entry: DiskStorageEntry? = null
        val file = getKeyFile(key)
        if (!file.exists()) {
            throw FileNotFoundException("DiskStorage.getEntryByKey() exists(): file doesn't exist ${key}")
        }
        var metadata: DiskStorageMetadata? = null
        var codec: Codec<*>? = null
        val metadataFile = getKeyMetadataFile(key)
        if (metadataFile.exists()) {
            try {
                metadata = DiskStorageMetadata.load(metadataFile)
                codec = if (metadata != null) {
                    getSerializer(key, metadata.entryClass)
                } else {
                    defaultCodec
                }
            } catch (e: Exception) {
                codec = defaultCodec
            }
        } else {
            codec = defaultCodec
        }
        if (codec == null) {
            throw Exception("Serializer is null")
        }
        entry = DiskStorageEntry(file, null, metadata, codec)
        return entry
    }

    @Throws(Exception::class)
    override fun remove(key: String) {
        val lockObject = getLockObject(key)
        synchronized(lockObject!!) {
            val entry = getEntry(key)
            entry?.delete()
        }
    }

    @Throws(Exception::class)
    override fun getEntries(): List<StorageEntry> {
        val entries: MutableList<StorageEntry> = ArrayList()
        if (!directory.exists()) {
            return entries
        }
        val files = directory.listFiles()
        if (files != null) {
            for (file in files) {
                if (!isMetadataFile(file)) {
                    val key = file.name
                    val lockObject = getLockObject(key)
                    synchronized(lockObject!!) {
                        val entry = getEntryByKey(key)
                        entries.add(entry)
                    }
                }
            }
        }
        return entries
    }

    val entryCount: Int
        get() {
            val files = directory.listFiles()
            var entryCount = 0
            for (file in files) {
                if (!isMetadataFile(file)) {
                    ++entryCount
                }
            }
            return entryCount
        }

    @Throws(Exception::class)
    override fun removeAll() {
        val entries = getEntries()
        for (file in entries) {
            val diskCacheEntry = file as DiskStorageEntry
            val key = diskCacheEntry.fileName
            val lockObject = getLockObject(key)
            synchronized(lockObject!!) { file.delete() }
        }
        if (!directory.delete()) {
            throw Exception("DiskStorage.removeAll() delete(): remove directory error")
        }
    }

    //// Setter
    fun setSerializer(codec: Codec<*>, cl: Class<*>) {
        classSerializerMap[cl] = codec
    }

    fun setSerializer(codec: Codec<*>, key: String) {
        keySerializerMap[key] = codec
    }

    fun setDefaultCodec(defaultCodec: Codec<*>) {
        this.defaultCodec = defaultCodec
    }

    companion object {
        private const val METADATA_PREFIX = "_metadata"
        private val ReservedCharsSet: Set<Char> = HashSet(Arrays.asList(
                '/', '|', '"', '?', '*', '<', '>', '\\', ':', '+', '[', ']'
        ))
    }

}